### ISSUE TITLE
FHS-like scheme for data

### ISSUE THREAD

**Issue by daler** [gogetdata/ggd-recipes#15]
Do you have a policy on which files should go where? I see that some recipes make subdirs in `$PREFIX/share/ggd/$species/$build`. For example `hg38-transcripts` goes in an `rnaseq` subdir, `hg38-bwa` makes a `bwa` subdir, but `hg38-hisat2` doesn't make a hisat2 dir. It seems like a lot of the ones with subdirs were copied from `../cloudbiolinux`.

Have you decided on some sort of filesystem hierarchy standard but for data?

**Comment by chapmanb** [comment]
Ryan;
This in imported from what we're doing with CloudBioLinux/bcbio right now. I don't know of any other conventions to follow, but we're using the $species/$build approach with subdirectories as you found. Generally, everything should go into subdirectories (I think hisat2 should unpack into one as well) and then we share some of the subdirectories like `variation` for things that come from multiple places and fit there. Happy to get suggestions or try to think up ways for new data. The general idea is to provide a sensible default so you can infer directory structure, then also be able to use ggd to list specific files for lookup.

**Comment by daler** [comment]
Thanks Brad.

(the reason I'm asking all of this is that I and collaborators are planning on adding stuff for dm6; trying to do things approximating the "right way" from the start)

 My guess is whatever you have in bcbio/CBL is probably the de facto standard! The only alternative I'm aware of is the [iGenomes]( http://support.illumina.com/sequencing/sequencing_software/igenome.html) scheme, which has directories along the lines of:
```
Homo_sapiens
└── UCSC
    └── hg38
        ├── Annotation
        │   ├── Genes
        │   ├── Genes.gencode
        │   └── SmallRNA
        └── Sequence
            ├── AbundantSequences
            ├── Bowtie2Index
            ├── BowtieIndex
            ├── BWAIndex
            │   ├── version0.5.x
            │   └── version0.6.0
            ├── Chromosomes
            └── WholeGenomeFasta
 ```
The useful bits here may be separating by provider (UCSC, NCBI, Ensembl) and having at least sequence and annotation dirs. I would think a separate indexes dir would be helpful too rather than nesting under sequences like they've done there. Then again, if all recipes are in a subdir according to their recipe name then a flat structure works too.

You're right, hisat2 unpacks into a subdir. I also checked the `hg38-transcripts` recipe, and there's all sorts of good stuff in there that gets unpacked. For future recipes, is the idea that something like `hg38-transcripts` be split into multiple recipes (i.e., a separate DEXSeq recipe, a separate tophat transcriptome index recipe, etc), or kept together in a bundle for a type of analysis (i.e. transcriptome)?

Last, you mention using ggd to list specific files. Is there currently a mechanism for that? I see from-bash and check-recipe subcommands. Would be awesome to have `ggd files hg38-hisat2` return the full paths of files created by that recipe, which would avoid the need for any kind of organizational decisions.

**Comment by arq5x** [comment]
Ryan and Brad, I agree that there should be a subdirectory structure, but I would like to give this a bit of thought and discussion, as, in my opinion, the current set of directories is a mix of "just right" and "too much" granularity. I think we should open a separate issue to discuss a) what directories make the most sense and b) the rules by which new subdirectories will be approved.

As for the `ggd files` command, I agree we need something like this, but @brentp just hasn't had time yet to work on that. We also want to develop the ability to cache prebuilt recipes so that when installed (`ggd install`?) it will pull from a cached location (S3?) by default and optionally build from scratch (`--build`).

**Comment by brentp** [comment]
I think that, yes, we should force that each recipe go into its own sub-directory. this will actually make the testing and tracking (e.g. for ggd files $recipe) much simpler. 

I'm thinking of scrapping all the automated conversions from bcbio. They'll be in the git history and should be easily updated 1-by-1, it's just a lot to keep doing when there are 70 of them.

**Comment by chapmanb** [comment]
Aaron, Brent and Ryan;
I'm not wedded to the current subdirectory structure so happy to hear suggestions. My only goal would that we could navigate and make sense of it without needing tooling. I'm not that good at thinking up specs so will happily migrate to what y'all come up with.

Once we've settled on something, I can transition bcbio/CloudBioLinux to use this instead of our current approach so we can maintain the recipes here and deprecate/remove the old ones. Happy to crank on this whenever ready. Thanks again for all the work on having a sweet infrastructure for this.

**Comment by brentp** [comment]
I think that we should have each recipe go to {species}/{build}/{name}/

where {name} is currently {genome_build}-{name}, e.g. `hg19-dbsnp` . Aaron suggested that we could also take an optional {source}, like ensembl or ncbi or ucsc. so `{genome_build}-{source}-{name}`

I like this as it keeps things simple and makes tracking very easy. I'll add to ggd check-recipe to make sure that a recipe only puts files in that sub-directory--without this, I just check what's modified and make sure it meets certain rules. This way, I can better track what is changed.

Of course users can choose to sym-link groups of recipes into their own hierarchy.

And, this will make `ggd path $recipe` quite simple.

Does this seem tolerable to everyone?

**Comment by arq5x** [comment]
I like this idea, as `source` could also things like bowtie, bedtools, smith-lab, etc.

One other complication I foresee is the potential confusion recipe users (not writers) will have in finding the precise name of a recipe that they want to install with `ggd path $recipe`, as the exact recipe name may not always be intuitive to them. 

Is it feasible to have something like `ggd search --species $species "genes"` that would return all potential recipes for that species that have the pattern either in the recipe name or contents?  That would enable other options like `ggd search --species $species --build $build "genes"`, etc.

What do you think?

**Comment by daler** [comment]
I like the simplified organization too.

Nomenclature-wise, I see the `source` in `{genome_build}-{source}-{name}` as referring to "source of original data" (e.g, ncbi, ensembl, smith_lab). In contrast, bowtie and bedtools might be better as part of a "to-be-used for" or "target audience" or "prepared by" field.

While `{genome_build}-{source}-{associated_tool}-{name}` gets to be unwieldy, is there a simpler way to differentiate between these:

- `mm10-ucsc-bowtie2-index`
- `mm10-ncbi-bowtie2-index`
- `mm10-ucsc-bwa-index`
- `mm10-ncbi-bwa-index`

or these:

- `hg38-gencode-source-annotation`
- `hg38-gencode-dexseq-annotation`
- `hg38-refseq-source-annotation`
- `hg38-refseq-dexseq-annotation`

?


Aaron's idea of `ggd search` for finding keywords in recipes would be really nice.

With respect to infrastructure and names, would it be possible for `ggd from-bash` to handle the creation of the proposed subdirs? I think it would be convenient if when writing a recipe we can assume the working directory for the script is `{species}/{build}/{name}`. Seems like there would be less room for error that way (like forgetting to move to $name).

**Comment by brentp** [comment]
quick response but will think about it more ...
great, that makes sense.

re `from-bash` handling the creation of sub-dirs, it already does that and then puts the bash script into the correct directory--the difference is that now the directory will be 1 level deeper.

in the CLI, we could add (yet another) option for adding a tag or tool (or whatever we call it) into the name.

**Comment by chapmanb** [comment]
Thanks for all the discussion. This all sounds great to me and happy to migrate to this. I do think there might be some value in de-emphasizing source in the directory structure. If this is included I'd have it be deeper in the tree. I'd also have sections be optional to make things cleaner.

My opinion is that we could purposely try to promote certain distributions to make things less complicated. For instance, in Ryan's example I'm not even sure I know what the difference is between these two:
```
mm10-ucsc-bowtie2-index
mm10-ncbi-bowtie2-index
```
Is that chr1 vs 1? If so, I found that a pain to support for GRCh37/hg19 and think we might be better served by having promoting one of them, say as:
```
mm10-bowtie
```
and if we have to:
```
mm10-bowtie-ncbi
```
I'd drop the index entirely as that seems like an extra unnecessary nesting.

In general we make things too unnecessarily complex in bioinformatics and just trying to retroactively get back the hours of my life wasted on chr1 versus 1 problems.

**Comment by brentp** [comment]
Good points. 

The chr1 vs 1 is determined entirely by the .genome file for that genome build. Even the current tests make sure that's the case.

I think in most cases, recipes will just be {genome_build}-{name} but the option could be there to have an additional (or 2?) specifier in the middle in cases where that makes sense.

I'm all for keeping things simple. We're flying to biology of genomes tomorrow so I will make a start of this then.

**Comment by daler** [comment]
Good points all. Yep, I was thinking of chr1 vs 1 in the ucsc vs ncbi case and avoiding the hours burned.
So given Brad's points, the annotation examples can be simplified to:

- `hg38-annotation`
- `hg38-annotation-dexseq`
- `hg38-annotation-refseq`
- `hg38-annotation-refseq-dexseq`

that is, gencode version is promoted to default and name gets more specific only as needed without needing any kind of a priori nesting.

**Comment by mikecormier** [comment]
Hey all. It has been a long time since this discussion. I just wanted to send out an update that these issues have been resolved. A subdirectory for recipes is based on {ggd channel}{species}{genome-build} and recipe names are based on {genome-build}-{name}-{data-provider}-{package-version}. 

chr or no chr prefix is determined by the genome build associated `.genome` file. For example, hg19 and hg38 recipe will have a chr prefix while GRCh37 and GRCh38 will not have a chr prefix. 

The cli has been updated to work with this directory/naming structure. The cli also provides the tooling for retrieving data file paths, listing installed data files, getting environment variables associated with installed data packages, and other management specific tools. 

Data packages can now be installed using `ggd install`. 

Additionally, multiple commands are equipped with a `--prefix` argument. This allows data packages to be installed into a different conda environment than the one currently being worked in, as well as accessing those data files from a different environment than where the data is stored. This minimizes duplicate installs of the same recipe littered across a single system and provides a way to access that data no matter what environment you are in (as long as ggd is installed in that environment.) NOTE: a few of the older recipes have not been updated to use the `--prefix` argument within `ggd install`, however, they will be updated at some point. 

There is a lot more we have added/updated. Many of the details can be found the ggd docs page: https://gogetdata.github.io/
